### ISSUE TITLE
twisted.conch.ssh.keys has poor tests.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2679]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| z3p reported|
|-|-|
|Trac ID|trac#2679|
|Type|defect|
|Created|2007-05-27 03:57:37Z|

The reasons are too large to enumerate, but twisted.conch.ssh.keys needs better tests.  Work is being done in the free-assert-2112-2 branch ATM.

<details><summary>Searchable metadata</summary>

```
trac-id__2679 2679
type__defect defect
reporter__z3p z3p
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__soc2007 soc2007
time__1180238257000000 1180238257000000
changetime__1194583842000000 1194583842000000
version__None None
owner__z3p z3p
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|z3p removed owner|
|-|-|

Work has been done on this in branches/free-assert-2112-2.  It is ready for review.
Note: Work has also been done on [#2679](https://github.com/twisted/twisted/issues/2679) for t.c.s.transport.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|z3p set owner to @therve|
|-|-|

Make that branches/keys-2679.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to z3p|
|-|-|

keys.py:
 * the change in getPublicKeyObject from 'data' to 'blob' as parameter name is somewhat gratuitous. If removed, you could revert changes on userauth and checkers.
 * the check at the beginning of getPublicKeyBlob should raise a meaningfull exception. It should be tested, too. The filename case and the unknow type of key is not tested either.
 * overall, almost no BadKeyError is tested
 * getPrivateKeyObject_agentv3 and makePrivateKeyString_agentv3  are not tested
 * printKey is not tested

test_keys.py:
 * HelpersTestCase overrides common.entropy.get_bytes but doesn't revert it in tearDown
 * test_signRSA and test_signDSA are not really tests, they should be prefixed by a '_'.
 * the case where Crypto is not present is not managed

General
 * Some pyflakes errors:
```
twisted/conch/ssh/userauth.py:19: 'MP' imported but unused
twisted/conch/ssh/userauth.py:90: undefined name 'ConchError'
twisted/conch/ssh/keys.py:16: 'string' imported but unused
twisted/conch/ssh/keys.py:26: 'log' imported but unused
twisted/conch/ssh/keys.py:356: undefined name 'keytype'
twisted/conch/checkers.py:23: 'IPluggableAuthenticationModules' imported but unused
```
2 of those errors show that test coverage must be improved :). (makePrivateKeyString_agentv3 is not called, the error case of ssh_USERAUTH_REQUEST is not tested).
 * the copyright notices of the touched files should be update to 2007
 * please remove trailing whitespaces
 * please use assertTrue instead of assert_ in the tests
 * default values of functions should be written ```foo='bar'``` instead of ```foo = 'bar'```
 * exceptions should be spelled ```raise Error('data')``` everywhere.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|z3p set owner to @therve|
|-|-|

keys.py:
* data -> blob: it may be gratuitous, but it helps avoid confusion with functions that can take data or a filename.  I'm okay with the two small tweaks to the other files.
* I've added tests for the other things

test_keys.py:
* I've refactored signDSA/RSA so that it's clear what's being tested.
* I've added the header which turns off the tests if Crypto isn't found.

pyflakes:
* I've fixed all of the pyflakes errors except the one in userauth.py.  There's an open ticket for that already, and it's outside the scope of this ticket.

other stuff:
* I think I've done all of them.

Putting this back in the review category.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to z3p|
|-|-|

There are still a few BadKeyErrors not tested, but that's OK. It also still fails if Crypto isn't found, but that's not a regression :).

I saw that you commented out some code in keys.py, I would prefer to see this removed if it's not useful (also, with no 'if 1' block).

Last thing, in setUp of SSHKeysHandlingTestCase, please use 'open' instead of 'file'.

Merging is OK for me, but that'd be cool if your mentor could have a quick look. Thanks.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|z3p set owner to @glyph|
|-|-|

OK, thanks for the help therve.  I'll bounce it to glyph for a check.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Just wanted to note that I'm aware this issue is waiting for some comment from me.  I've been extremely overcommitted recently, but I will be getting to this by this weekend at the latest.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

keys.py could probably use some refactoring to simplify key interactions in conch.  I'd suggest having a class which represents a key (maybe a class for public and a class for private) with class methods for loading from different forms and methods for serializing to different formats.

Some other comments about keys.py:

  * sibling imports should be replaced with absolute imports
  * the string after the sibling imports should be a comment or part of the module docstring.  Aggregating key interaction functionality into a class would probably obviate the need to convey this information distinctly, since the class's methods would describe the different serializations possible, and instance of the class would themselves replace direct use of Crypto objects.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Of course I was including monday in the "weekend".

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph set owner to z3p|
|-|-|

Here are some things that should be fixed before merging anything:

  * The data -> blob change really is gratuitous.  I don't like the fact that there are changes to ```twisted.conch.checkers``` which don't come with an accompanying change in a ```test_checkers```  - and ```test_keys``` doesn't import ```test_checkers```.  Normally I wouldn't object to argument name changes, but the fact that previous internal suggested usage for this (publicly named) function was to call by keyword arguments implies that there is potentially external code which would need to go through a deprecation cycle to safely change this.  (See [the draft compatibility policy](https://twistedmatrix.com/trac/wiki/CompatibilityPolicy) for details.)
  * There's a spurious "(" added to twisted.conch.ssh.keys' docstring.
  * ```getPublicKeyString``` needs a docstring, at least indicating that it's deprecated.
    * similarly for a bunch of other functions being changed.
  * trailing whitespace added in ```_testKey```'s docstring.

However, there are some other things that should probably be completed before this ticket is considered done.

First of all, there's still some uncovered code in ```keys.py```.
  * ```getPublicKeyString```, though deprecated, should still be covered. 
  * There are no negative-path tests which cover ```makePublicKeyString_lsh```, ```makePrivateKeyString_lsh```, ```makePrivateKeyString_openssh```, and ```makePublicKeyString_openssh```'s exception-raising paths.
  * ```getPrivateKeyObject_agentv3``` and ```makePrivateKeyString_lsh``` are never invoked such that p is greater than q, so the swap branch is never tested.

I also strongly second exarkun's comment about having an "object-oriented" interface to this functionality.  I wish I could say that twisted.internet._sslverify provided a good example, but it's unfortunately completely undocumented.  Have a look anyway, maybe the structure there will make intuitive sense to you.

It would also be nice to have finer-grained tests for certain things that would be difficult to test given the current factoring.  For example, ```getPrivateKeyObject``` doesn't separate its key-type-detection logic from directly invoking the loaders for the various kinds of data.

That type-detection code, as well as things like ```objectType``` are subjected only to "drive-by" testing, being tested deep in the middle of some other functionality.  It would be good to have more isolated testing of the individual bits, so if future maintainers break anything, they'll be able to more readily identify which part of the code is the problem.

I understand the need for static data in tests dealing with static data, but it would be nice if it could be packed in such a way that it would fit the coding standard's <80 chars rule.  And perhaps it should be in a specific fixtures file so that it wouldn't detract from the logic of the test cases themselves.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|z3p removed owner|
|-|-|

I've refactored all those functions into a class (Key), deprecated all the functions, and added tests for some of the functions that weren't tested explicitly before.  I also broke the key data into it's own file (test/keydata.py), which accounts for the changes to test_ssh and test_conch.
I've looked for the other things you both mentioned, but with the refactoring I'm not sure I've gotten them.  I'm placing it back up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to z3p|
|-|-|

there are a couple test failures and a debug print in test_keys.py, it'd be nice to fix the wildcard import there too (so pyflakes can work, and to comply w/ coding standard)

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|z3p set owner to @exarkun|
|-|-|

Okay, not sure why I was being so stupid there, but it's fixed now.  Bouncing it back.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to z3p|
|-|-|

  * twisted/conch/ssh/keys.py
    * There's a spurious "(" added to twisted.conch.ssh.keys' docstring.
    * Don't use the @ decorator syntax - it doesn't work in Python 2.3
    * in _fromString_public_openssh et al, the public_openssh part should be capitalized to indicate dynamic dispatch.
    * The _fromString_* methods should have docstrings, including parameter and return types and raised exceptions
    * ```object``` isn't the most awesome name - maybe something that indicates that it is a PyCrypto key object of some sort?
    * !__eq!__ will most likely raise an exception if a Key is ever compared to anything that isn't a Key.  The most correct thing for it to do in that case would be to return NotImplemented.  An almost-as-correct thing would be to return False.
    * Why try/bare-except in !__repr!__?
    * Same comments about _toString_* as _fromString_*
    * Deprecated functions should use warnings.warn() to emit DeprecationWarnings (with unit tests - see TestCase.assertWarns)
    * No deprecations in minor releases - if there is a version number in the deprecation message, it should be 0.9.0
  * twisted/conch/ssh/common.py
    * didn't I see this change somewhere else already?

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|z3p removed owner|
|-|-|

I think I've addressed these issues.  Re: t.c.ssh.common, I checked my branches and I don't see it anywhere else.  Ready for review in branches/keys-2679-2.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

Urr, I was gonna review this, but ran out of time.  The one thing I did notice though, is that in keys.py, Key.fromFile ignores its type and passphrase arguments.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|z3p set owner to @exarkun|
|-|-|

I've fixed this in the branch.  Marking it for exarkun.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to z3p|
|-|-|

It looks like ```Key.fromFile``` is still ignoring its type and passphrase arguments.

```__eq__``` is still not right.  Returning NotImplemented is different from raising NotImplementedError (<http://jcalderone.livejournal.com/32837.html>).  There should be tests for this code, I guess.

I don't think assertWarns will work properly if an exception is raised out of the function it is given, so WarningsTestCase might need to have its assertions split up to separate the tests for the warnings from the tests for the exceptions.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_19"></a>|z3p set owner to @exarkun|
|-|-|

I'm not sure what you mean by #2; the code does raise NotImplementedError, the test was wrong.  Ready for review in keys-2679-2.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set owner to z3p|
|-|-|

  * common.py
    * getMP used to return a two-tuple of an int and a str.  Now it returns a two-tuple of a tuple and a str.  There don't seem to be any tests for getMP, and no existing tests are broken by this change.  I suppose the code isn't even executed by the test suite, since the gmpy version is generally used.  I see a lot of callers of getMP, and none seem to account for the API change.  I suspect this change shouldn't be made.  A new function can be introduced if some code needs this functionality (but since nothing is expecting it, I'm not sure what might need it).
  * keys.py
    * some trailing whitespace
    * invalid epytext markup in makePublicKeyString for the `kind' parameter
    * objectType function shouldn't use ```except:```
    * missing whitespace around operators in pkcs1Pad and pkcs1Digest
  * test_keys.py
    * SSHKeysHandlingTestCase and HelpersTestCase have class-level deprecation suppressions which are too broad

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

btw, epytext errors:

```
  L10: Error: Unbalanced '{'.

       C{str
        ^
   L3: Error: Improper paragraph indentation.
   L3: Warning: Possible mal-formatted field item.
   L3: Error: Improper paragraph indentation.
   L4: Error: Improper paragraph indentation.
   L7: Error: Improper paragraph indentation.
   L8: Error: Improper paragraph indentation.
   L5: Error: Improper paragraph indentation.
these 5 objects' docstrings are not proper epytext:
    twisted.conch.ssh.keys.makePublicKeyString
    twisted.conch.ssh.keys.Key._fromString_PRIVATE_LSH
    twisted.conch.ssh.keys.Key.data
    twisted.conch.ssh.keys.Key.blob
    twisted.conch.ssh.keys.Key.toString
```

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_22"></a>|z3p removed owner|
|-|-|

getMP does not now return a two tuple of a list and string; it returns a variable-length tuple of count+1 strings.  I've updated _fastgetMP to have the same signature.  I've fixed the other items as well.  Back up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun set owner to z3p|
|-|-|

Seems in good shape now.  The only issue which remains that I would draw attention to is the amount of code in conch itself which is now using deprecated APIs.  I can think of three things which might be done about that:

  1. change the code in question to use non-deprecated APIs in this branch
  1. file a ticket for making that change
  1. use PendingDeprecationWarning instead of DeprecationWarning and worry about it later

Your choice.  Aside from this, I don't think anything is blocking a merge.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_24"></a>|z3p set status to `closed`|
|-|-|

(In [21097]) refactor t.c.ssh.keys.* into an object and rewrite tests

Author: z3p
Reviewers: exarkun, glyph, therve
Fixes [#2679](https://github.com/twisted/twisted/issues/2679)

twisted.conch.ssh.keys.* was previously a collection of functions that were
poorly documented and poorly tested.  This refactors that module's
functionality into a single class: Key.  The old functions remain, but are
deprecated.  The tests in test/test_keys have been updated to include the
refactoring as well as increasing the test coverage.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_25"></a>|z3p commented|
|-|-|

(In [21100]) Back out 21097 (win32 test failures)

Refs [#2679](https://github.com/twisted/twisted/issues/2679)

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_26"></a>|z3p set status to `reopened`|
|-|-|

Test failures on win32

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_27"></a>|z3p set owner to @exarkun<br>z3p set status to `new`|
|-|-|

Fixed test failures on win32.  Back up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun set owner to @glyph|
|-|-|

Hey, review this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_30"></a>|@exarkun set owner to z3p<br>@exarkun set status to `new`|
|-|-|

Tests pass on win32-py2.5-select now.  win32-py2.4-select has been offline for a few days.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_31"></a>|z3p set owner to @exarkun|
|-|-|

There were conflict when I tried to merge this, due to the new randbytes module.  I've made the changes in keys-2679-3, and am putting this back up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_32"></a>|@exarkun set owner to z3p|
|-|-|

There was already a keys-2679-3 branch.  I filed a ticket for the Combinator bug that lets it copy trunk into an existing branch when you run ```mkbranch``` and deleted trunk from ```/branches/keys-2679-3```.  There weren't any changes in it, and apparently the existing branch was new enough to already have the randbytes changes, so it merges cleanly to trunk anyway.  I had to copy keydata.py from the -2 branch though, it got missed in the merge forward I guess.  Then, I couldn't run the branch on some of the builders because of the recent incompatible trial change, so I merged the whole thing forward again to **keys-2679-4**.

  * twisted/conch/test/test_ssh.py - update copyright date
  * twisted/conch/test/test_keys.py - test_RSA's docstring refers to testDSA, but that test method is called test_DSA now.

I added some trivial whitespace to a couple places.  Otherwise the branch looks good.  Merge it once you fix the two points above.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_33"></a>|z3p set status to `closed`|
|-|-|

(In [21670]) Refactor t.c.ssh.keys.* into an object and rewrite tests

Author: z3p
Reviewers: exarkun, glyph, therve
Fixes [#2679](https://github.com/twisted/twisted/issues/2679)

twisted.conch.ssh.keys.* was previously a collection of functions that were
poorly documented and poorly tested. This refactors that module's functionality
into a single class: Key. The old functions remain, but are deprecated. The 
tests in test/test_keys have been updated to include the refactoring as well as
increasing the test coverage.
